File: pyref/atlasopt/graph.py
"""data.json からのグラフ構築。

- mastery ノード(グループ中央の飾り、辺なし)は除外する
- データ上の擬似ノード "root" は捨て、スタートノード 29045 を root として扱う
- root のコストは 0、他は全ノード 1(Atlas Tree の仕様)
- 辺は in/out を統合して無向として扱う
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

ROOT_ID = "29045"  # スタートノード(名前空)。ポイントを消費しない

_PSEUDO_ROOT = "root"

_REPO_ROOT = Path(__file__).resolve().parents[2]
# 標準は submodule(vendor/atlastree-export)。ルート直下のベタ置きはフォールバック
_DATA_CANDIDATES = (
    _REPO_ROOT / "vendor" / "atlastree-export" / "data.json",
    _REPO_ROOT / "data.json",
)


def _default_data() -> Path:
    for p in _DATA_CANDIDATES:
        if p.is_file():
            return p
    raise FileNotFoundError(
        "data.json not found — run `git submodule update --init` "
        f"(searched: {', '.join(map(str, _DATA_CANDIDATES))})"
    )


@dataclass(frozen=True)
class AtlasGraph:
    adj: dict[str, frozenset[str]]
    info: dict[str, dict]  # 生のノード情報(mastery・擬似rootを除く)
    mastery_notables: dict[str, tuple[str, ...]]  # masteryノードid -> 同グループのNotable id
    root: str = ROOT_ID

    def cost(self, node: str) -> int:
        return 0 if node == self.root else 1

    @property
    def nodes(self) -> frozenset[str]:
        return frozenset(self.adj)


def load(path: str | Path | None = None) -> AtlasGraph:
    data_path = Path(path) if path else _default_data()
    raw = json.loads(data_path.read_text())
    nodes = raw["nodes"]

    def is_graph_node(nid: str) -> bool:
        return nid != _PSEUDO_ROOT and not nodes[nid].get("isMastery")

    adj: dict[str, set[str]] = {n: set() for n in nodes if is_graph_node(n)}
    for nid, nd in nodes.items():
        if not is_graph_node(nid):
            continue
        for other in list(nd.get("out", [])) + list(nd.get("in", [])):
            if other != nid and other in adj:
                adj[nid].add(other)
                adj[other].add(nid)

    if ROOT_ID not in adj:
        raise ValueError(f"start node {ROOT_ID} not found in {data_path}")

    notables_by_group: dict[int, list[str]] = {}
    for nid, nd in nodes.items():
        if nid != _PSEUDO_ROOT and nd.get("isNotable"):
            notables_by_group.setdefault(nd["group"], []).append(nid)
    mastery_notables = {
        nid: tuple(notables_by_group.get(nd["group"], ()))
        for nid, nd in nodes.items()
        if nid != _PSEUDO_ROOT and nd.get("isMastery")
    }

    info = {nid: nodes[nid] for nid in adj}
    return AtlasGraph(
        adj={n: frozenset(s) for n, s in adj.items()},
        info=info,
        mastery_notables=mastery_notables,
    )


def reachable_from(adj: dict[str, frozenset[str]], start: str, banned: frozenset[str] = frozenset()) -> set[str]:
    """banned を通らずに start から到達できるノード集合(start含む)。"""
    if start in banned or start not in adj:
        return set()
    seen = {start}
    stack = [start]
    while stack:
        cur = stack.pop()
        for nxt in adj[cur]:
            if nxt not in seen and nxt not in banned:
                seen.add(nxt)
                stack.append(nxt)
    return seen

File: web/src/data/graph.ts
/**
 * data.json からのグラフ構築(pyref/atlasopt/graph.py の移植)。
 *
 * - mastery ノード(グループ中央の飾り、辺なし)は除外する
 * - データ上の擬似ノード "root" は捨て、スタートノード 29045 を root として扱う
 * - root のコストは 0、他は全ノード 1(Atlas Tree の仕様)
 * - 辺は in/out を統合して無向として扱う
 *
 * pyref と違いファイルI/Oは持たない。パース済み JSON を受け取る純関数のみ
 * (Node のテストは tests/helpers/data.ts、ブラウザは fetch(M4)がI/Oを担当)。
 */

export const ROOT_ID = "29045"; // スタートノード(名前空)。ポイントを消費しない

const PSEUDO_ROOT = "root";

export interface AtlasNode {
  skill?: number;
  name?: string;
  icon?: string;
  stats?: string[];
  reminderText?: string[];
  flavourText?: string[];
  group?: number;
  orbit?: number;
  orbitIndex?: number;
  isMastery?: boolean;
  isNotable?: boolean;
  isKeystone?: boolean;
  isWormhole?: boolean;
  grantedPassivePoints?: number;
  in?: string[];
  out?: string[];
  [key: string]: unknown;
}

export interface AtlasGroup {
  x: number;
  y: number;
  orbits: number[]; // 信頼できない(実ノードと食い違うグループがある)。描画ヒント程度
  nodes: string[];
  background?: { image: string; offsetX?: number; offsetY?: number };
}

export interface SpriteCoord {
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface SpriteSheet {
  filename: string; // CDN URL。basename がローカル assets/ のファイル名
  coords: Record<string, SpriteCoord>;
}

export interface AtlasData {
  nodes: Record<string, AtlasNode>;
  groups?: Record<string, AtlasGroup>;
  constants?: {
    skillsPerOrbit: number[];
    orbitRadii: number[];
    [key: string]: unknown;
  };
  /** カテゴリ名 → ズームキー(文字列) → シート。ズームキーは imageZoomLevels でなくここから導出する */
  sprites?: Record<string, Record<string, SpriteSheet>>;
  points?: { totalPoints?: number };
  [key: string]: unknown;
}

export interface AtlasGraph {
  readonly adj: ReadonlyMap<string, ReadonlySet<string>>;
  readonly info: ReadonlyMap<string, AtlasNode>; // 生のノード情報(mastery・擬似rootを除く)
  readonly masteryNotables: ReadonlyMap<string, readonly string[]>; // masteryノードid -> 同グループのNotable id
  readonly root: string;
}

export function cost(g: AtlasGraph, node: string): number {
  return node === g.root ? 0 : 1;
}

export function buildGraph(data: AtlasData): AtlasGraph {
  const nodes = data.nodes;

  const isGraphNode = (nid: string): boolean =>
    nid !== PSEUDO_ROOT && !nodes[nid]?.isMastery;

  const adj = new Map<string, Set<string>>();
  for (const nid of Object.keys(nodes)) {
    if (isGraphNode(nid)) adj.set(nid, new Set());
  }
  for (const [nid, nd] of Object.entries(nodes)) {
    if (!isGraphNode(nid)) continue;
    for (const other of [...(nd.out ?? []), ...(nd.in ?? [])]) {
      if (other !== nid && adj.has(other)) {
        adj.get(nid)!.add(other);
        adj.get(other)!.add(nid);
      }
    }
  }

  if (!adj.has(ROOT_ID)) {
    throw new Error(`start node ${ROOT_ID} not found in data`);
  }

  const notablesByGroup = new Map<number, string[]>();
  for (const [nid, nd] of Object.entries(nodes)) {
    if (nid !== PSEUDO_ROOT && nd.isNotable && nd.group !== undefined) {
      let bucket = notablesByGroup.get(nd.group);
      if (!bucket) {
        bucket = [];
        notablesByGroup.set(nd.group, bucket);
      }
      bucket.push(nid);
    }
  }
  const masteryNotables = new Map<string, readonly string[]>();
  for (const [nid, nd] of Object.entries(nodes)) {
    if (nid !== PSEUDO_ROOT && nd.isMastery) {
      masteryNotables.set(
        nid,
        nd.group !== undefined ? (notablesByGroup.get(nd.group) ?? []) : [],
      );
    }
  }

  const info = new Map<string, AtlasNode>();
  for (const nid of adj.keys()) info.set(nid, nodes[nid]!);

  return { adj, info, masteryNotables, root: ROOT_ID };
}

/** banned を通らずに start から到達できるノード集合(start含む)。反復DFS。 */
export function reachableFrom(
  adj: ReadonlyMap<string, ReadonlySet<string>>,
  start: string,
  banned: ReadonlySet<string> = new Set(),
): Set<string> {
  if (banned.has(start) || !adj.has(start)) return new Set();
  const seen = new Set([start]);
  const stack = [start];
  while (stack.length > 0) {
    const cur = stack.pop()!;
    for (const nxt of adj.get(cur)!) {
      if (!seen.has(nxt) && !banned.has(nxt)) {
        seen.add(nxt);
        stack.push(nxt);
      }
    }
  }
  return seen;
}
